File: CS-Examples/02_FindAndReplace/ReplaceWithHtml.cs
using System;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.IO;
using Spire.Doc;
using Spire.Doc.Documents;
using Spire.Doc.Fields;
using System.Collections.Generic;

namespace ReplaceWithHtml
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            // Read HTML content from a file and store it in a String variable
            String HTML = File.ReadAllText(@"..\..\..\..\..\..\Data\InputHtml1.txt");

            // Create a new Document object from a template file and store it in a Document variable
            Document document = new Document(@"..\..\..\..\..\..\Data\ReplaceWithHtml.docx");

            // Create a new List to store DocumentObject objects
            List<DocumentObject> replacement = new List<DocumentObject>();

            // Add a new section to the document and store it in a Section variable
            Section tempSection = document.AddSection();

            // Add a new paragraph to the section and store it in a Paragraph variable
            Paragraph par = tempSection.AddParagraph();
            // Append the HTML content to the paragraph
            par.AppendHTML(HTML);

            // Iterate through all child objects of the section
            foreach (DocumentObject obj in tempSection.Body.ChildObjects)
            {
                // Cast the current object to DocumentObject and add it to the replacement list
                DocumentObject docObj = obj as DocumentObject;
                replacement.Add(docObj);
            }

            // Find all occurrences of the string "[#placeholder]" in the document and store them in an array of TextSelection objects
            TextSelection[] selections = document.FindAllString("[#placeholder]", false, true);

            // Create a new List to store TextRangeLocation objects
            List<TextRangeLocation> locations = new List<TextRangeLocation>();
            // Iterate through all TextSelection objects found
            foreach (TextSelection selection in selections)
            {
                // Get the range of the current selection and create a new TextRangeLocation object with it
                locations.Add(new TextRangeLocation(selection.GetAsOneRange()));
            }
            // Sort the locations list in ascending order
            locations.Sort();

            // Iterate through all TextRangeLocation objects in the locations list
            foreach (TextRangeLocation location in locations)
            {
                // Call the ReplaceWithHTML method with the current location and replacement list as arguments
                ReplaceWithHTML(location, replacement);
            }

            // Remove the temporary section added earlier from the document
            document.Sections.Remove(tempSection);

            // Save the modified document to a file named "Output.docx" with the Docx file format
            document.SaveToFile("Output.docx", FileFormat.Docx);
            // Dispose of the document object to release resources
            document.Dispose();

            //Launch the Word file.
            WordDocViewer("Output.docx");
        }

        private static void ReplaceWithHTML(TextRangeLocation location, List<DocumentObject> replacement)
        {
            TextRange textRange = location.Text;

            //textRange index
            int index = location.Index;

            //get owner paragraph
            Paragraph paragraph = location.Owner;

            //get owner text body
            Body sectionBody = paragraph.OwnerTextBody;

            //get the index of paragraph in section
            int paragraphIndex = sectionBody.ChildObjects.IndexOf(paragraph);

            int replacementIndex = -1;
            if (index == 0)
            {
                //remove the first child object
                paragraph.ChildObjects.RemoveAt(0);

                replacementIndex = sectionBody.ChildObjects.IndexOf(paragraph);
            }
            else if (index == paragraph.ChildObjects.Count - 1)
            {
                paragraph.ChildObjects.RemoveAt(index);
                replacementIndex = paragraphIndex + 1;
            }
            else
            {
                //split owner paragraph
                Paragraph paragraph1 = (Paragraph)paragraph.Clone();
                while (paragraph.ChildObjects.Count > index)
                {
                    paragraph.ChildObjects.RemoveAt(index);
                }
                int i = 0;
                int count = index + 1;
                while (i < count)
                {
                    paragraph1.ChildObjects.RemoveAt(0);
                    i += 1;
                }
                sectionBody.ChildObjects.Insert(paragraphIndex + 1, paragraph1);

                replacementIndex = paragraphIndex + 1;
            }

            //insert replacement
            for (int i = 0; i <= replacement.Count - 1; i++)
            {
                sectionBody.ChildObjects.Insert(replacementIndex + i, replacement[i].Clone());
            }
        }

        public class TextRangeLocation : IComparable<TextRangeLocation>
        {
            public TextRangeLocation(TextRange text)
            {
                this.Text = text;
            }

            public TextRange Text
            {
                get { return m_Text; }
                set { m_Text = value; }
            }

            private TextRange m_Text;
            public Paragraph Owner
            {
                get { return this.Text.OwnerParagraph; }
            }

            public int Index
            {
                get { return this.Owner.ChildObjects.IndexOf(this.Text); }
            }

            public int CompareTo(TextRangeLocation other)
            {
                return -(this.Index - other.Index);
            }
        }

        private void WordDocViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }
    }
}

File: VB-Examples/02_FindAndReplace/ReplaceWithHtml.vb
Imports System.Text
Imports System.IO
Imports Spire.Doc
Imports Spire.Doc.Documents
Imports Spire.Doc.Fields

Namespace ReplaceWithHtml
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			' Read the contents of the input HTML file into a string.
			Dim HTML As String = File.ReadAllText("..\..\..\..\..\..\Data\InputHtml1.txt")

			' Create a new Document object and load a Word document to replace with HTML.
			Dim document As New Document("..\..\..\..\..\..\Data\ReplaceWithHtml.docx")

			' Create a list to store the replacement HTML content as DocumentObjects.
			Dim replacement As New List(Of DocumentObject)()

			' Create a temporary section in the document.
			Dim tempSection As Section = document.AddSection()

			' Add a paragraph to the temporary section and append the HTML content.
			Dim par As Paragraph = tempSection.AddParagraph()
			par.AppendHTML(HTML)

			' Retrieve the DocumentObjects from the temporary section and add them to the replacement list.
			For Each obj As DocumentObject In tempSection.Body.ChildObjects
				Dim docObj As DocumentObject = TryCast(obj, DocumentObject)
				replacement.Add(docObj)
			Next obj

			' Find all occurrences of the placeholder "[#placeholder]" in the document.
			Dim selections() As TextSelection = document.FindAllString("[#placeholder]", False, True)

			' Create a list to store the locations of the text range containing the placeholder.
			Dim locations As New List(Of TextRangeLocation)()
			For Each selection As TextSelection In selections
				locations.Add(New TextRangeLocation(selection.GetAsOneRange()))
			Next selection

			' Sort the locations based on the index of the text range.
			locations.Sort()

			' Iterate through each location and replace the text range with the HTML content.
			For Each location As TextRangeLocation In locations
				ReplaceWithHTML(location, replacement)
			Next location

			' Remove the temporary section from the document.
			document.Sections.Remove(tempSection)

			' Save the modified document to the specified output file in Docx format.
			document.SaveToFile("Output.docx", FileFormat.Docx)

			' Dispose of the Document object to release resources.
			document.Dispose()
			
			'Launch the Word file.
			WordDocViewer("Output.docx")
		End Sub

		' Method to replace a text range with HTML content.
		Private Sub ReplaceWithHTML(ByVal location As TextRangeLocation, ByVal replacement As List(Of DocumentObject))
			Dim textRange As TextRange = location.Text
			Dim index As Integer = location.Index
			Dim paragraph As Paragraph = location.Owner
			Dim sectionBody As Body = paragraph.OwnerTextBody
			Dim paragraphIndex As Integer = sectionBody.ChildObjects.IndexOf(paragraph)
			Dim replacementIndex As Integer = -1

			If index = 0 Then
				' Remove the first child object (text range) from the paragraph.
				paragraph.ChildObjects.RemoveAt(0)
				replacementIndex = sectionBody.ChildObjects.IndexOf(paragraph)
			ElseIf index = paragraph.ChildObjects.Count - 1 Then
				' Remove the last child object (text range) from the paragraph.
				paragraph.ChildObjects.RemoveAt(index)
				replacementIndex = paragraphIndex + 1
			Else
				' Clone the current paragraph and split its child objects before and after the text range.
				Dim paragraph1 As Paragraph = CType(paragraph.Clone(), Paragraph)
				Do While paragraph.ChildObjects.Count > index
					paragraph.ChildObjects.RemoveAt(index)
				Loop
				Dim i As Integer = 0
				Dim count As Integer = index + 1
				Do While i < count
					paragraph1.ChildObjects.RemoveAt(0)
					i += 1
				Loop
				' Insert the cloned paragraph after the original paragraph within the body.
				sectionBody.ChildObjects.Insert(paragraphIndex + 1, paragraph1)
				replacementIndex = paragraphIndex + 1
			End If

			' Insert the replacement HTML content at the specified index within the body.
			For i As Integer = 0 To replacement.Count - 1
				sectionBody.ChildObjects.Insert(replacementIndex + i, replacement(i).Clone())
			Next i
		End Sub

		' Class representing the location of a text range within a paragraph.
		Public Class TextRangeLocation
			Implements IComparable(Of TextRangeLocation)

			Public Sub New(ByVal text As TextRange)
				Me.Text = text
			End Sub

			' Property for the TextRange object.
			Public Property Text() As TextRange
				Get
					Return m_Text
				End Get
				Set(ByVal value As TextRange)
					m_Text = value
				End Set
			End Property

			Private m_Text As TextRange

			' Property for retrieving the owner paragraph of the text range.
			Public ReadOnly Property Owner() As Paragraph
				Get
					Return Me.Text.OwnerParagraph
				End Get
			End Property

			' Property for retrieving the index of the text range within its owner paragraph.
			Public ReadOnly Property Index() As Integer
				Get
					Return Me.Owner.ChildObjects.IndexOf(Me.Text)
				End Get
			End Property

			' Implementation of the CompareTo method for sorting purposes.
			Public Function CompareTo(ByVal other As TextRangeLocation) As Integer Implements IComparable(Of TextRangeLocation).CompareTo
				' Compare two text range locations based on their indices (descending order).
				Return -(Me.Index - other.Index)
			End Function
		End Class

		Private Sub WordDocViewer(ByVal fileName As String)
			Try
				Process.Start(fileName)
			Catch
			End Try
		End Sub
	End Class
End Namespace
